libs/libc/netdb/lib_dnsaddserver.c

```diff
@@ -87,7 +87,7 @@ int dns_add_nameserver(FAR const struct sockaddr *addr, socklen_t addrlen)
   int status;
   int ret;
 
-  stream = fopen(CONFIG_NETDB_RESOLVCONF_PATH, "at");
+  stream = fopen(CONFIG_NETDB_RESOLVCONF_PATH, "r");
```

Comment:
Just change to r is incorrect, the code intend to update etc/resolv.conf at line 183 and 189.

Reply:
In that case, I think the original author intended "a+" for "append/update." However, there is a bigger problem because as Surya points out, "We have experimented with "a+" "w+" etc but these did not work because of the "resolv.conf" file resides in /etc directory in ROMFS, ROMFS supports only "r" mode." [1](https://lists.apache.org/thread.html/rbbbee93f79dd8f4224969bb03bf093d8c96421380a1d955167a2f661%40%3Cdev.nuttx.apache.org%3E) So a correct fix will likely require a larger design change.

Reply:
If user enable CONFIG_NETDB_RESOLVCONF, he/she should put resolv.conf in the writable filesystem by CONFIG_NETDB_RESOLVCONF_PATH.
Surya report two problem:
1.Call open with "at": this is a typo bug we need fix DNS(change to "r+").
2. Fail to open with "a+"/"w+" due to read-only ROMFS: this is invalid usage, Surya should modify his config.

Reply:
i guess the original code meant "a".
some fopen implementations recognize "t" as "text mode".

Reply:
Since it is /etc/resolv.conf is a read-only file, none of those explanations make sense.  I would write it off as just a dumb typo.  The code was obviously never used.

Reply:
> 
> 
> If user enable CONFIG_NETDB_RESOLVCONF, he/she should put resolv.conf in the writable filesystem by CONFIG_NETDB_RESOLVCONF_PATH.
> Surya report two problems:
> 1.Call open with "at": this is a typo bug we need fix DNS(change to "a+").
> 2. Fail to open with "a+"/"w+" due to read-only ROMFS: this is invalid usage, Surya should modify his config.

Based on Xaio Xang's comment, the "r" needs to be changed to "a+".  There are still uncorrected checks that also need to be addressed.

Reply:
Perhaps there should be an option to support a read-only /etc/resolv.conf?

Reply:
> Perhaps there should be an option to support a read-only /etc/resolv.conf?

But if you read the whole function carefully, you can find dns_add_nameserver don't has any other real action if the file can't write.

Reply:
> i guess the original code meant "a".
> some fopen implementations recognize "t" as "text mode".

Yes, @yamt your analysis is right, but "t" isn't specified in standard:
https://pubs.opengroup.org/onlinepubs/009695399/functions/fopen.html
And since dns_add_nameserver never read from the opened stream, so the better fix is changed to "a".

Reply:
> But if you read the whole function carefully, you can find dns_add_nameserver don't has any other real action if the file can't write.

From the little reading I have done, that does not seem exactly true:

- If DHCP is used the, yes, /etc/resolve/conf may be updated.  In this case, resolv.conf would need to be on writable media.  But updates to resovle.conf from DHCP is not supported by our implementation of DHCP.
- On static IP configurations without DHCP, resolve.conf is not modified by the system, and hence would be effectly read-only.

Or am I reading that wrong.

Reply:
> If the above is correct, then at the present time the argument should be "r" but would have to be revisited if the capability of DHCPC were improved.

Hmm.. but clearly I am mistaken because libs/libc/netdb/lib_dnsaddserver.c does, indeed, write to /etc/resolv.conf